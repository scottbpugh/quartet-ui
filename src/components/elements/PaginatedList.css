.paginated-list-table{
    width:100%;
}
