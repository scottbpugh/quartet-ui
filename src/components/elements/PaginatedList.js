// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {Card, Tag, ControlGroup, Button, InputGroup,
  Intent} from "@blueprintjs/core";
import {FormattedMessage} from "react-intl";
import {withRouter} from "react-router";
import {pluginRegistry} from "plugins/pluginRegistration";

/*
  Displays a list of objects (entries, events, companies, locations) as
  single cards.
*/
class _PaginatedList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filter: "",
      keywordSearch: "",
      entries: [],
      entriesPerPage: 20,
      inputSize: 50,
      maxPages: 1,
      count: 0,
      interactive: 'pt-interactive',
      loading: true,
      loadingRR: true,
      currentPage: null,
    };
    this.offset = 0;
    this.currentPage = 1;
    this.debounced = null;
    this.fetchEntries = null;
    this.timer = null;
  }

  // filter by a field in the rows.
  filterBy = evt => {
    this.setState({filter: evt.currentTarget.value}, () => {
      this.offset = 0;
      this.currentPage = 1;
    });
  };

  // search by a field in the rows or all of them.
  searchBy = evt => {
    // this.setState({keywordSearch: evt.currentTarget.value}, () => {
      this.offset = 0;
      this.currentPage = 1;
      this.processEntries();
    // });
  };
  updateSearch = evt => {
    this.setState({keywordSearch: evt.currentTarget.value});
  };
  handleEnterKeySearch = evt => {
    if (evt.key === "Enter") {
      this.searchBy(evt);
    }
  };
  componentDidMount() {
    if (sessionStorage.getItem("loadingRR") === null ) {
      sessionStorage.setItem("loadingRR", false);
    }
    this.processEntries();
    this.setState({
      entries: this.props.entries,
      maxPages: 1,
      count: 0,
    });
    this.timer = setInterval(()=> {
      if(sessionStorage.getItem("loading") != this.state.loading || sessionStorage.getItem("loadingRR") != this.state.loadingRR) {
        this.setState({
          loading: JSON.parse(sessionStorage.getItem("loading")),
          loadingRR: JSON.parse(sessionStorage.getItem("loadingRR")),
        });
        
      };
      if (sessionStorage.getItem("loadingRR") != this.state.loadingRR) {
        this.setState({
          loading: JSON.parse(sessionStorage.getItem("loading")),
          loadingRR: JSON.parse(sessionStorage.getItem("loadingRR")),
        });
      }
      if (this.state.loadingRR === null) {
        this.setState({
          loadingRR: false
        })
      }
    }
    , 5);
  }
  componentWillUnmount() {
    clearInterval(this.timer);
  }
  // refresh the Lists, keeping the search filters.
  componentWillReceiveProps(nextProps) {
    let maxPages = this.currentPage;
    if (nextProps.next !== null && Array.isArray(nextProps.entries)) {
      maxPages = Math.ceil(nextProps.count / nextProps.entries.length);
      // console.log("nextProps.count", nextProps.count)
      // console.log("nextProps.entries.length", nextProps.entries.length)
      // console.log(Math.ceil(nextProps.count / nextProps.entries.length));
      if(Number.isNaN(maxPages) || maxPages === Infinity) {
        // console.log("Change state: ", Number.isNaN(maxPages));
        
      }
      // if(maxPages === Infinity) {
      //   this.setState({
      //     maxPages: 1
      //   });
      // }
      else {
        this.setState({
          maxPages: maxPages
        });
      }
    }
    this.setState({
      currentPage: this.currentPage
    })
    this.setState({
      entries: nextProps.entries,
      count: nextProps.count,
      interactive: this.props.interactive === false ? '' : ' pt-interactive',
      loading: JSON.parse(sessionStorage.getItem("loading")),
      loadingRR: JSON.parse(sessionStorage.getItem("loadingRR"))
    });
    
  }

  goTo = path => {
    this.props.history.push(path);
  };

  // go to next page if possible.
  next = () => {
    this.currentPage += 1;
    this.processEntries(true);
  };

  // go to previous page if possible.
  previous = () => {
    this.currentPage -= 1;
    this.processEntries(true);
  };

  loadingScreen = () => {
    this.setState(
      { loading : true },
      () => {
          setTimeout(()=>{this.setState({loading : false})}, 6000);
      }
    );
  };


  processEntries = (clear = false) => {
    if (this.debounced) {
      clearTimeout(this.debounced);
    }
    this.debounced = setTimeout(() => {
      const {loadEntries, server} = this.props;
      loadEntries(
        server,
        this.state.keywordSearch,
        this.currentPage,
        this.props.ordering ? this.props.ordering : null,
        this.props.type ? this.props.type : null
      );
    }, clear);
    // this.loadingScreen(); 
  };

  render() {
    // console.log(this.state.maxPages)
    const {entries} = this.state;
    return (
      <Card className="pt-elevation-4">
        <h5>
          {" "}
          <div className="right-aligned-elem">
            <Button
              style={{marginRight: "5px"}}
              title="refresh"
              iconName="refresh"
              onClick={this.processEntries.bind(this, true)}
            />
            <Tag className="pt-large">
              {this.currentPage}/{this.state.maxPages}
            </Tag>
          </div>
          {this.props.listTitle ? this.props.listTitle : null}
        </h5>
        <div>
          <div className="table-control">
            <div className="pagination-control">
              <div>
                <Button
                  disabled={this.currentPage - 1 < 1 || this.state.loading === true || this.state.loadingRR === true}
                  onClick={this.previous.bind(this)}>
                  previous
                </Button>{" "}
                |{" "}
                <Button
                  disabled={this.currentPage >= this.state.maxPages  || this.state.loading === true || this.state.loadingRR === true}
                  onClick={this.next.bind(this)}>
                  next
                </Button>
              </div>
            </div>
            <div>
              <ControlGroup fill={false} vertical={false}>
              <Button intent={Intent.PRIMARY} onClick={this.searchBy}>Search</Button>
                {/* <div className="pt-select">
                  <select value={this.state.filter}>
                    <option value="">Search</option>
                  </select>
                </div> */}
                <InputGroup
                  onChange={this.updateSearch}
                  onKeyPress={this.handleEnterKeySearch}
                  value={this.state.keywordSearch}
                  placeholder={pluginRegistry
                    .getIntl()
                    .formatMessage({id: "app.common.enterKeywords"})}
                />
              </ControlGroup>
              <div className="label-info-display">
                <FormattedMessage
                  id="app.common.entriesTotal"
                  values={{entriesCount: this.state.count}}
                />
              </div>
            </div>
          </div>
          <div className="overflowed-table">
            <table className={`pt-table pt-bordered pt-striped ${this.state.interactive}`}>
              <this.props.tableHeaderClass server={this.props.server} />
              <tbody
                style={{
                  textAlign: "center",
                  verticalAlign: "middle !important"
                }}>
                {Array.isArray(entries) && entries.length > 0 && this.state.loading === false && this.state.loadingRR === false && this.currentPage===this.state.currentPage
                  ? entries.map((entry, index) => {
                      return (
                        <this.props.entryClass
                          key={`entry-${index}`}
                          entry={entry}
                          loadEntries={this.props.loadEntries}
                          server={this.props.server}
                          history={this.props.history}
                          page={this.currentPage}
                        />
                      );
                    })
                    : 
                    this.state.loading === true 
                    || 
                    this.state.loadingRR === true ?
                  <tr className='tableLoading'>
                    <div class="middle">
                        <div class="bar bar1"></div>
                        <div class="bar bar2"></div>
                        <div class="bar bar3"></div>
                        <div class="bar bar4"></div>
                        <div class="bar bar5"></div>
                        <div class="bar bar6"></div>
                        <div class="bar bar7"></div>
                        <div class="bar bar8"></div>
                    </div>
                  </tr>  
                  : 
                  this.state.keywordSearch != "" && entries != undefined && entries.length === 0 && this.state.loading === false && this.state.loadingRR === false && this.currentPage===this.state.currentPage?
                  <tr className='tableLoading'>
                      <div class="middle searchResult">
                          <FormattedMessage
                              id="app.common.searchResult"
                              defaultMessage="No search result"
                          />
                      </div>
                  </tr>
                  : this.state.keywordSearch === "" && entries != undefined && entries.length === 0 && this.state.loading === false && this.state.loadingRR === false && this.currentPage===this.state.currentPage?
                  <tr className='tableLoading'>
                      <div class="middle searchResult">
                      <FormattedMessage
                              id="app.common.emptyArray"
                              defaultMessage="Empty array"
                          />
                      </div>
                  </tr>
                  :
                  <tr className='tableLoading'>
                    <div class="middle">
                        <div class="bar bar1"></div>
                        <div class="bar bar2"></div>
                        <div class="bar bar3"></div>
                        <div class="bar bar4"></div>
                        <div class="bar bar5"></div>
                        <div class="bar bar6"></div>
                        <div class="bar bar7"></div>
                        <div class="bar bar8"></div>
                    </div>
                  </tr> 
                  }
              </tbody>
            </table>
          </div>
        </div>
      </Card>
    );
  }
}

export const PaginatedList = withRouter(_PaginatedList);

window.qu4rtet.exports("components/elements/PaginatedList", this);
