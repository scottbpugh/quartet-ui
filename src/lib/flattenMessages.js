// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export function flattenMessages(nestedMessages, prefix = "") {
  return Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;

    if (typeof value === "string") {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }

    return messages;
  }, {});
}

window.qu4rtet.exports("lib/flattenMessages", this);
