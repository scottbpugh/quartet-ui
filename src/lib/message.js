// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Position, Toaster, Intent} from "@blueprintjs/core";
import {pluginRegistry} from "plugins/pluginRegistration";
import stringHash from "string-hash";

const MINIMUM_EXPIRE = 7000;

const msgToaster = Toaster.create({
  className: "my-toaster",
  position: Position.BOTTOM_RIGHT
});

const getIntent = type => {
  switch (type) {
    case "success":
      return Intent.SUCCESS;
    case "danger":
    case "error":
      return Intent.DANGER;
    case "warning":
      return Intent.WARNING;
    default:
      return Intent.PRIMARY;
  }
};

const _msgHashMap = {};

const showMessageNoRepeat = (toastMessage, expires_in = MINIMUM_EXPIRE) => {
  let msgHash = stringHash(JSON.stringify(toastMessage));
  let timestamp = new Date().getTime();
  if (msgHash in _msgHashMap && _msgHashMap[msgHash].expires > timestamp) {
    // don't show msg and update the expiration to prevent the same message
    // over and over.
    _msgHashMap[msgHash].expires = timestamp + expires_in;
    return;
  }
  _msgHashMap[msgHash] = {expires: timestamp + expires_in};
  msgToaster.show(toastMessage);
};

export const showMessage = msg => {
  try {
    // for errors etc out of a component context.
    const intl = pluginRegistry.getIntl();
    if (msg.id) {
      showMessageNoRepeat(
        {
          message: intl.formatMessage({id: msg.id}, msg.values),
          intent: getIntent(msg.type)
        },
        msg.expires_in || MINIMUM_EXPIRE
      );
    } else if (msg.msg) {
      //let formatMsg = typeof msg.msg === "object" ? JSON.stringify(msg.msg) : msg.msg;
      showMessageNoRepeat(
        {message: msg.msg, intent: getIntent(msg.type)},
        msg.expires_in || MINIMUM_EXPIRE
      );
    }
  } catch (e) {
    console.log(
      "an error occurred while attempting to display an error toast.",
      e
    );
  }
};

window.qu4rtet.exports("lib/message", this);
