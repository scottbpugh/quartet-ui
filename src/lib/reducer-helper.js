// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// A helper function meant to make life a little easier in reducers.

/**
 * Unknown - Description
 *
 * @param {object} [state={}]       The reducer state
 * @param {type}   serverID         The serverID
 * @param {object} [mergeObject={}] The object to merge into the state for this server.
 *
 * @return {type} Description
 */
export const setServerState = (state = {}, serverID, mergeObject = {}) => {
  if (!serverID) {
    throw new Error(
      "A server ID must be defined to set the server state in reducer."
    );
  }
  const copy = {...state}; // use a copy of state for manipulations.
  if (!copy.servers) {
    // if servers are not created yet, create an object.
    copy.servers = {};
  }
  if (!copy.servers[serverID]) {
    // if there is no object for this server, create it.
    copy.servers[serverID] = {};
  }
  return {
    ...copy,
    servers: {
      ...copy.servers,
      [serverID]: {
        ...copy.servers[serverID],
        ...mergeObject
      }
    }
  };
};
window.qu4rtet.exports("lib/reducer-helper", this);
