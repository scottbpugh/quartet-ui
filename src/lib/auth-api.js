// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers.
 */
export const prepHeaders = (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Description
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      let postFields = data.actions.POST;
      let formStructure = Object.keys(postFields)
        .map(field => {
          if (postFields[field].read_only === false) {
            return {name: field, description: postFields[field]};
          } else {
            return null;
          }
        })
        .filter(fieldObj => {
          if (processField) {
            return processField(fieldObj);
          }
          if (fieldObj) {
            // create sync validation arrays.
            fieldObj.validate = getSyncValidators(fieldObj);
            return true;
          }
          return false;
        });
      createForm(formStructure);
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};
