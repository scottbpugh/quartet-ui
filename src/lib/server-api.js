// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeadersAuth - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  const headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method,
    headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Returns an object with a form structure to be used by PageForm/redux form
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      try {
        const postFields = data.actions.POST;
        const formStructure = Object.keys(postFields)
          .map(field => {
            if (
              field === "id" ||
              (postFields[field].type === "field" &&
                postFields[field].read_only === true)
            ) {
              return null;
            }
            return {name: field, description: postFields[field]};
          })
          .filter(fieldObj => {
            if (processField) {
              return processField(fieldObj);
            }
            if (fieldObj) {
              // create sync validation arrays.
              fieldObj.validate = getSyncValidators(fieldObj);
              return true;
            }
            return false;
          });
        createForm(formStructure);
      } catch (e) {
        if (e.message.startsWith("Cannot read property 'POST'")) {
          pluginRegistry.getHistory().push("/access-denied");
        }
      }
    })
    .catch(error => {
      debugger;
      console.log("About to redirect to access denied");
      //pluginRegistry.getHistory().push("/access-denied");
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

/**
 * fetchObject - fetches a single object from any Swagger API endpoint.
 *
 * @param {Server}   serverInstance  A Server instance
 * @param {string} [operationId=]  The Swagger operation id
 * @param {object} [parameters={}] Parameter passed with the call.
 *
 * @return {object} An object from the API call.
 */
export const fetchObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchPageList - Fetches a single list page.
 *
 * @param {type} serverInstance A Server instance
 * @param {type} operationId    The Swagger
 * @param {type} parameters     Parameters passed to the call (page...)
 *
 * @return {object} A response body
 */
export const fetchPageList = async (
  serverInstance,
  operationId,
  parameters
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
    throw new Error(response);
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchListAll - Loops through all pages of a list and concats the result.
 *
 * @param {Server}   serverInstance  A Server Instance
 * @param {string} [operationId=]  Swagger operation id
 * @param {object} [parameters={}] Parameters passed to the call.
 * @param {array}  [results=[]]    The results, all objects for this API list.
 *
 * @return {type} Description
 */
export const fetchListAll = async (
  serverInstance,
  operationId = "",
  parameters = {},
  results = []
) => {
  return new Promise((resolve, reject) => {
    serverInstance
      .getClient()
      .then(client => {
        client
          .execute({
            operationId,
            parameters,
            securities: {
              authorized: client.securities,
              specSecurity: [client.spec.securityDefinitions]
            }
          })
          .then(response => {
            if (response.ok) {
              if (Array.isArray(response.body)) {
                results = results.concat(response.body);
                resolve(results);
              } else if (response.body.results) {
                results = results.concat(response.body.results);
                // pagination in effect. Assuming page number navigation.
                if (response.body.next) {
                  const url = new URL(response.body.next);
                  const page = new URLSearchParams(url.search).get("page");
                  const subParameters = {...parameters, page};
                  fetchListAll(operationId, subParameters, results)
                    .then(resolve)
                    .catch(reject);
                } else {
                  resolve(results);
                }
              }
            } else {
              reject(response);
            }
          })
          .catch(e => {
            reject(e);
          });
      })
      .catch(e => {
        reject(e);
      });
  });
};

/**
 * deleteObject - Deletes an object.
 *
 * @param {string} [operationId=]  The operationId to execute.
 * @param {object} [parameters={}] Object data
 *
 * @return {object} response.
 */
export const deleteObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  const response = await client.execute({
    operationId,
    parameters,
    securities: {
      authorized: client.securities,
      specSecurity: [client.spec.securityDefinitions]
    }
  });
  if (response.ok) {
    return response;
  }
};

window.qu4rtet.exports("lib/server-api", this);
