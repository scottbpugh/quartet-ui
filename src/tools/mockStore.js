// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React from "react";

import {IntlProvider} from "react-intl";
import {addLocaleData} from "react-intl";
import messages from "messages";
import en from "react-intl/locale-data/en";
import fr from "react-intl/locale-data/fr";
import {flattenMessages} from "lib/flattenMessages";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

addLocaleData([...en, ...fr]);
let defaultLocale = "en-US";
const middlewares = [thunk];

export const initialState = {
  dashboard: {notifications: []},
  serversettings: {
    servers: {
      "824ae2dd-2411-4adf-a846-d9b310a33d4b": {
        serverSettingName: "Server #1",
        serverName: "localhost",
        port: "80",
        ssl: false,
        username: "root",
        password: "toor",
        serverID: "824ae2dd-2411-4adf-a846-d9b310a33d4b"
      }
    }
  },
  intl: {
    defaultLocale: defaultLocale,
    locale: defaultLocale,
    messages: flattenMessages(messages[defaultLocale])
  },
  layout: {pageTitle: {id: "app.nav.servers"}},
  plugins: {plugins: {}, navTreeItems: []}
};

export const mockStore = configureStore(middlewares);

export const TestWrapper = ({locale, messages, children}) => {
  if (locale === undefined) {
    locale = defaultLocale;
  }
  if (messages === undefined) {
    messages = initialState.intl.messages;
  }
  return (
    <IntlProvider locale={locale} messages={messages}>
      {children}
    </IntlProvider>
  );
};

