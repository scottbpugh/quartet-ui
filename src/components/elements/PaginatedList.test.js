// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import {TestWrapper} from "tools/mockStore";
import Enzyme, {mount, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";
import toJson from "enzyme-to-json";
import * as serverAPI from "lib/server-api";
import {PaginatedList} from "./PaginatedList";

const TestTableHeader = props => (
  <thead>
    <th>
A Single Value
    </th>
  </thead>
);

const TestEntry = props => (
  <tr key={props.entry.key}>
    <td>
      {props.entry.name}
    </td>
  </tr>
);

it("renders correctly", () => {
  const list = renderer
    .create(
      <TestWrapper>
        <PaginatedList
          listTitle="Hello Test"
          loadEntries={() => {}}
          server={{serverID: "blabla"}}
          entries={[{name: "Hello World", key: "hello"}]}
          count={0}
          next={0}
          tableHeaderClass={TestTableHeader}
          entryClass={TestEntry}
        />
      </TestWrapper>
    )
    .toJSON();
  expect(list).toMatchSnapshot();
});
