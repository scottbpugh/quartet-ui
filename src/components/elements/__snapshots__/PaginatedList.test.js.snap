// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="pt-card pt-elevation-0 pt-elevation-4"
  >
    <h5>
       
      <div
        className="right-aligned-elem"
      >
        <button
          className="pt-button pt-icon-refresh"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            Object {
              "marginRight": "5px",
            }
          }
          title="refresh"
          type="button"
        />
        <span
          className="pt-tag pt-large"
        >
          1
          /
          1
        </span>
      </div>
      Hello Test
    </h5>
    <div>
      <div
        className="table-control"
      >
        <div
          className="pagination-control"
        >
          <div>
            <button
              className="pt-button pt-disabled"
              disabled={true}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="button"
            >
              <span>
                previous
              </span>
            </button>
             
            |
             
            <button
              className="pt-button pt-disabled"
              disabled={true}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="button"
            >
              <span>
                next
              </span>
            </button>
          </div>
        </div>
        <div>
          <div
            className="pt-control-group"
          >
            <div
              className="pt-select"
            >
              <select
                value=""
              >
                <option
                  value=""
                >
                  Search
                </option>
              </select>
            </div>
            <div
              className="pt-input-group"
            >
              <input
                className="pt-input"
                onChange={[Function]}
                placeholder="Enter Keywords..."
                style={
                  Object {
                    "paddingRight": 0,
                  }
                }
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="label-info-display"
          >
            <span>
              0 Entries Total
            </span>
          </div>
        </div>
      </div>
      <div
        className="overflowed-table"
      >
        <table
          className="pt-table pt-bordered pt-striped pt-interactive"
        >
          <thead>
            <th>
              A Single Value
            </th>
          </thead>
          <tbody
            style={
              Object {
                "textAlign": "center",
                "verticalAlign": "middle !important",
              }
            }
          >
            <tr>
              <td>
                Hello World
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
